### DISCUSSION TITLE
[WINUI3] Is SwapChainPanel's Opacity property supported at all?

### DISCUSSION THREAD

**Discussion by hereafter** [microsoft/microsoft-ui-xaml#8191] [Q&A] [answered]
I have issues to make a transparent/semi-transparent SwapChainPanel work.

I tried the following 

```
<SwapChainPanel Opacity="0.5">
</SwapChainPanel >

```

The content is opaque always.

And when i composite

```
<Grid Opacity="0.5">
<SwapChainPanel>
</SwapChainPanel>
</Grid>

```
The grid becomes opaque too.

What are the correct ways to make a SwapChainPanel transparent?

The content of the swap chain panel is rendered with Direct2D with a target of the following descriptions

```

        DXGI_SWAP_CHAIN_DESC1 swapChainDesc{ 0 };
        swapChainDesc.Width = width;
        swapChainDesc.Height = height;
        swapChainDesc.Format = DXGI_FORMAT_B8G8R8A8_UNORM;
        swapChainDesc.Stereo = false;
        swapChainDesc.SampleDesc.Count = 1;
        swapChainDesc.SampleDesc.Quality = 0;
        swapChainDesc.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
        swapChainDesc.BufferCount = 2;
        swapChainDesc.Scaling = DXGI_SCALING_STRETCH;
        swapChainDesc.AlphaMode = DXGI_ALPHA_MODE_PREMULTIPLIED;
        swapChainDesc.SwapEffect = DXGI_SWAP_EFFECT_FLIP_SEQUENTIAL;
        swapChainDesc.Flags = 0;
```

**Comment by castorix** [comment]
They added a black/white background (depending on current theme) in Windows App SDK 1.1.0
It can be removed with _SetLayeredWindowAttributes_ like in the test I had posted ([WinUI3_SwapChainPanel_Layered](https://github.com/castorix/WinUI3_SwapChainPanel_Layered))

**Reply by hereafter** [reply]
interesting I am not looking for ways to make irregular window transparency which can be archived by use direct composition.

I would like a swap chain panel to blend with its content behind , from my app as well . Normal xaml hi element has a properly opacity, as if it was ignored in my testing.

Is this a bug or anything I am missing

**Comment by castorix** [comment]
Opacity works, but as I said, the SwapChainPanel is covered by the background...

**Reply by hereafter** [reply]
I have issues to get it work.

I cloned your demo,

and made the following changes
```

<Window
    x:Class="WinUI3_SwapChainPanel_Layered.MainWindow"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:WinUI3_SwapChainPanel_Layered"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d">
    <Grid Background="Red">
        <SwapChainPanel x:Name="swapChainPanel1" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" Opacity="0.1">
            <Border x:Name="mainBorder" BorderThickness="2" CornerRadius="45" BorderBrush="Red" Background="Transparent" Margin="5, 5, 5, 5">
                <StackPanel Orientation="Vertical" HorizontalAlignment="Stretch" VerticalAlignment="Stretch">
                    <TextBlock x:Name="tb1" Text="Right-click to exit" Margin="5,5,5,5" HorizontalAlignment="Center" VerticalAlignment="Center" Foreground="Yellow"/>
                    <Button x:Name="myButton" Click="myButton_Click" HorizontalAlignment="Center" VerticalAlignment="Center" Margin="10, 140, 10, 10" Background="Blue">
                        Set Bitmap
                    </Button>
                </StackPanel>
            </Border>
        </SwapChainPanel>
    </Grid>
</Window>
```


Basic idea, behind this is there is some content behind the SwapChainPanel from the same window, it doesn't blend.
After running the Red Grid is ignored.

I would like the swapchainpanel to behave like all other xaml element nodes

**Comment by castorix** [comment]
> After running the Red Grid is ignored.
> I would like the swapchainpanel to behave like all other xaml element nodes

If you change for example the width :  _Width ="400"_  of SwapChainPanel , you will see the Red Grid.
But the SwapChainPanel is a bit different from other XAML controls : it has no Background (if you set Background in XAML, it will crash)

**Reply by hereafter** [reply]
thanks for clarifying.

Back to question shouldn't SwapChainPanel control honor its opacity setting 0.5 
so i can see through partially revealing its backdrop red grid content?

Is this behavior a known limitation, or is it a bug?
any workaround?

**Comment by castorix** [comment]
Not sure of the result you want to get, but a workaround can be to subclass with WM_ERASEBKGND and use the same color as parent Grid
If I do a test similar as [Windows with C++ : High-Performance Window Layering Using the Windows Composition Engine](https://learn.microsoft.com/en-us/archive/msdn-magazine/2014/june/windows-with-c-high-performance-window-layering-using-the-windows-composition-engine)
with Blue circle with Yellow background and 0.5f Alpha

Normal :
![SwapChainPanel_Normal](https://user-images.githubusercontent.com/22345506/220095530-a6f4b1be-b04c-4811-84ce-19b332aa5908.jpg)

No Background :
![SwapChainPanel_NoBackground](https://user-images.githubusercontent.com/22345506/220095564-e713d7f0-0139-46c6-97c3-138631bc5db9.jpg)

With Background :
![SwapChainPanel_Background](https://user-images.githubusercontent.com/22345506/220095604-caa76e96-30d3-488d-a343-dfeda81993f0.jpg)

**Reply by hereafter** [reply]
interesting, i am exploring the api sets. Not necessary a solid background.
I will keep the limitation and workaround in mind thanks

**Comment by lhak** [comment, marked as answer]
Unfortunately, blending a swapchain with winui 3 content is not currently supported (see https://github.com/microsoft/microsoft-ui-xaml/issues/6893). This is a regression compared to the inbox XAML layer.

**Reply by MikeHillberg** [reply]
See also note in the Remarks of the [doc page](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/Microsoft.UI.Xaml.Controls.SwapChainPanel?view=windows-app-sdk-1.2#remarks)

**Reply by nbryant42** [reply]
Transparency works in limited ways, but it has to be done at pixel level in the swapchain's alpha channel.

**Comment by axojhf** [comment]
I've faced a similar issue where I want to display some animation effects by overlaying a SwapChainPanel on top of existing WinUI3 controls.
I'm curious how you solve similar problems.

**Comment by zkvii** [comment]
I dont know whether this is what u want [transparent swapchainpanel](https://github.com/zkvii/WinUI_transparent_swapchain)
